### PR TITLE
Initial support for CompositeFilter class

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java [hunk 1]
```diff
@@ -553,44 +603,71 @@ private void validateOrderByFieldMatchesInequality(
...
+    for (FieldFilter subfilter : filter.getAllFieldFilters()) {
+      validateAddingNewFieldFilter(testQuery, subfilter);
+      testQuery = query.filter(subfilter);
+    }
+  }
+
+  /**
+   * Checks if any of the provided filter operators are included in the given list of filters and
+   * returns the first one that is, or null if none are.
+   */
```
Comment (ehsannas) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java:658]: This function was moved from firestore/core/Query.java to firestore/Query.java as it was the only consumer of it.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/Filter.java [hunk 2]
```diff
@@ -15,11 +15,19 @@
...
 
 public abstract class Filter {
   /** Returns true if a document matches the filter. */
   public abstract boolean matches(Document doc);
 
   /** A unique ID identifying the filter; used when serializing queries. */
   public abstract String getCanonicalId();
+
+  /** Returns a list of all field filters that are contained within this filter */
+  public abstract List<FieldFilter> getAllFieldFilters();
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/Filter.java:29]: It's not very obvious to me what the difference between "getFilters" in CompositeFIlter and and "getAllFieldFilters" is. Should it be getFIlters/getFlattenedFilters. As it stands now, getAllFieldFilters() suggests that it is is merely `getFilters().filter(f -> f instanceof FieldFilter)`.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/Filter.java [hunk 3]
```diff
@@ -15,11 +15,19 @@
...
+
+  /** Returns the field of the first filter that's an inequality, or null if none. */
+  public abstract FieldPath inequalityField();
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/Filter.java:32]: Please prefix with "get" (and possibly future-proof it by making it `getFirstInequalityField`)

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  public boolean isDisjunction() {
+    // TODO(orquery): Replace with Operator.OR.
+    return operator == Operator.OPERATOR_UNSPECIFIED;
+  }
+
+  /**
+   * Performs a depth-first search to find and return the first FieldFilter in the composite filter
+   * that satisfies the condition. Returns null if none of the FieldFilters satisfy the condition.
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java:75]:
```suggestion
   * that satisfies the condition. Returns {@code null} if none of the FieldFilters satisfy the condition.
```

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+   */
+  private FieldFilter firstFieldFilterWhere(Function<FieldFilter, Boolean> condition) {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java:77]:
```suggestion
  @Nullable private FieldFilter firstFieldFilterWhere(Function<FieldFilter, Boolean> condition) {
```

Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java:77]: Also, just because I can't help myself and to give you something to ignore: What about "findFirstMatchingFilter"? This method name could use a verb.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    StringBuilder builder = new StringBuilder();
+    for (Filter filter : filters) {
+      if (builder.length() == 0) {
+        builder.append(isConjunction() ? "and(" : "or(");
+      } else {
+        builder.append(",");
+      }
+      builder.append(filter.getCanonicalId());
+    }
+    builder.append(")");
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java:125]: You can use `TextUtils.join()`:

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/FieldFilter.java [hunk 7]
```diff
@@ -158,6 +159,20 @@ public String getCanonicalId() {
     return getField().canonicalString() + getOperator().toString() + Values.canonicalId(getValue());
   }
 
+  @Override
+  public List<FieldFilter> getAllFieldFilters() {
+    // This is already a field filter, so we return a list of size one.
+    return Arrays.asList(this);
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/FieldFilter.java:165]:
```suggestion
    return Collections.singletonList(this);
```

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java [hunk 8]
```diff
@@ -426,15 +429,62 @@ private FieldFilter parseFieldFilter(@NonNull FieldPath fieldPath, Operator op,
...
+    List<com.google.firebase.firestore.core.Filter> parsedFilters = new ArrayList<>();
+    for (Filter filter : compositeFilterData.getFilters()) {
+      com.google.firebase.firestore.core.Filter parsedFilter = parseFilterData(filter);
+      if (parsedFilter != null) {
+        parsedFilters.add(parsedFilter);
+      }
+    }
+    // For empty composite filters, return null.
+    // For example: and(or(), and()).
+    if (parsedFilters.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java]: Done

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java [hunk 9]
```diff
@@ -426,15 +429,62 @@ private FieldFilter parseFieldFilter(@NonNull FieldPath fieldPath, Operator op,
...
   // TODO(orquery): This method will become public API. Change visibility and add documentation.
   private Query where(Filter filter) {
-    return new Query(
-        query.filter(parseFieldFilter(filter.getField(), filter.getOperator(), filter.getValue())),
-        firestore);
+    com.google.firebase.firestore.core.Filter parsedFilter = parseFilterData(filter);
+    if (parsedFilter == null) {
+      // Return the existing query if not adding any more filters (e.g. an empty composite filter).
+      return this;
+    }
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java]: Done

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java [hunk 10]
```diff
@@ -426,15 +429,62 @@ private FieldFilter parseFieldFilter(@NonNull FieldPath fieldPath, Operator op,
...
+   */
+  @Nullable
+  private com.google.firebase.firestore.core.Filter parseFilterData(Filter filter) {
+    hardAssert(
+        filter instanceof FieldFilterData || filter instanceof CompositeFilterData,
+        "Parsing is only supported for FieldFilterData and CompositeFilterData.");
+    if (filter instanceof FieldFilterData) {
+      return parseFieldFilterData((FieldFilterData) filter);
+    }
+    return parseCompositeFilterData((CompositeFilterData) filter);
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java]: I like the idea because it'll get rid of the `instanceof` checking here. But that'd require passing both `FirebaseFirestore` object and the `core/Query` object to the `Filter.parse()` method.

 the signature would have to be the following:
```java
public com.google.firebase.firestore.core.Filter parse(FirebaseFirestore firestore, com.google.firebase.firestore.core.Query query);
```
because the parsing logic needs these.

not sure if it's worth it.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Filter.java [hunk 11]
```diff
@@ -147,6 +122,66 @@ public static Filter notInArray(@NonNull String field, @Nullable Object value) {
...
+  @NonNull
+  public static Filter and(Filter... filters) {
+    return new CompositeFilterData(
+        Arrays.asList(filters), StructuredQuery.CompositeFilter.Operator.AND);
+  }
+}
+
+/** @hide */
+@RestrictTo(RestrictTo.Scope.LIBRARY)
+class FieldFilterData extends Filter {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/Filter.java]: There is also already a FieldFilter class. Between two FieldFilters and two UnaryFilters, I would pick two UnaryFilters as  the naming UnaryFilter vs CompositeFilter offers clear and obvious distinctions.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Filter.java [hunk 12]
```diff
@@ -17,37 +17,60 @@
...
-    this.operator = operator;
-    this.value = value;
-  }
-
   /** @hide */
   @RestrictTo(RestrictTo.Scope.LIBRARY)
-  public FieldPath getField() {
-    return field;
-  }
+  static class FieldFilter extends Filter {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/Filter.java:30]: This is not public and thus should not need the annotation. Same below.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java [hunk 13]
```diff
@@ -426,15 +428,56 @@ private FieldFilter parseFieldFilter(@NonNull FieldPath fieldPath, Operator op,
               .parseQueryValue(value, op == Operator.IN || op == Operator.NOT_IN);
     }
     FieldFilter filter = FieldFilter.create(fieldPath.getInternalPath(), op, fieldValue);
-    validateNewFilter(filter);
     return filter;
   }
 
+  /**
+   * Takes a {@link Filter.CompositeFilter} object, parses each of its subfilters, and a new {@link
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java:435]: Word missing before "a new". The return type in the comment is also not correct.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java [hunk 14]
```diff
@@ -426,15 +428,56 @@ private FieldFilter parseFieldFilter(@NonNull FieldPath fieldPath, Operator op,
...
+   * CompositeFilter} that is constructed using the parsed values. Returns null if the given
+   * Filter.CompositeFilter does not contain any subfilters. Returns a {@link FieldFilter} if the
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java:437]: Method no longer returns null.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java [hunk 15]
```diff
@@ -426,15 +428,56 @@ private FieldFilter parseFieldFilter(@NonNull FieldPath fieldPath, Operator op,
...
+    return parseCompositeFilter((Filter.CompositeFilter) filter);
+  }
+
   // TODO(orquery): This method will become public API. Change visibility and add documentation.
   private Query where(Filter filter) {
-    return new Query(
-        query.filter(parseFieldFilter(filter.getField(), filter.getOperator(), filter.getValue())),
-        firestore);
+    com.google.firebase.firestore.core.Filter parsedFilter = parseFilter(filter);
+    if (parsedFilter.getFlattenedFilters().isEmpty()) {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java:475]: Would it be more efficient to call getFilters() here?

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java [hunk 16]
```diff
@@ -553,44 +596,71 @@ private void validateOrderByFieldMatchesInequality(
...
+      testQuery = query.filter(subfilter);
+    }
+  }
+
+  /**
+   * Checks if any of the provided filter operators are included in the given list of filters and
+   * returns the first one that is, or null if none are.
+   */
+  @Nullable
+  private Operator findFilterOperator(
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java:653]: s/findFilterOperator/findFilterByOperarator or findFilterWithOperator?

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+    return filters;
+  }
+
+  public Operator getOperator() {
+    return operator;
+  }
+
+  @Override
+  public List<FieldFilter> getFlattenedFilters() {
+    List<FieldFilter> result = new ArrayList<>();
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java:46]: Would it make sense to memoize this result?

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+    builder.append(isConjunction() ? "and(" : "or(");
+    TextUtils.join(",", canonicalIds);
+    builder.append(")");
+    return builder.toString();
+  }
+
+  @Override
+  public String toString() {
+    return getCanonicalId();
+  }
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java:132]: We probably also need equals() and hashCode()

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/Filter.java [hunk 19]
```diff
@@ -15,11 +15,19 @@
...
   public abstract boolean matches(Document doc);
 
   /** A unique ID identifying the filter; used when serializing queries. */
   public abstract String getCanonicalId();
+
+  /** Returns a list of all field filters that are contained within this filter */
+  public abstract List<FieldFilter> getFlattenedFilters();
+
+  /** Returns the field of the first filter that's an inequality, or null if none. */
+  public abstract FieldPath getFirstInequalityField();
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/Filter.java:32]: Add `@Nullable`.

**Review by schmidt-sebastian — COMMENTED** [review]
Some more minor comments. Thanks for the update.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Filter.java [hunk 20]
```diff
@@ -147,6 +122,66 @@ public static Filter notInArray(@NonNull String field, @Nullable Object value) {
...
+  }
+
+  public Object getValue() {
+    return value;
+  }
+}
+
+/** @hide */
+@RestrictTo(RestrictTo.Scope.LIBRARY)
+class CompositeFilterData extends Filter {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/Filter.java]: Leaving as is is fine.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/Filter.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/Filter.java]: Pinging this once more as I find the current naming a bit hard to follow. It is not obvious to me that a CompositeFilter is not a FieldFilter.

[the remark this answers is not in the document]
Comment (ehsannas) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/Filter.java]: Done

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+  public List<FieldFilter> getFlattenedFilters() {
+    List<FieldFilter> result = new ArrayList<>();
+    for (Filter subfilter : filters) {
+      result.addAll(subfilter.getFlattenedFilters());
+    }
+    return result;
+  }
+
+  /**
+   * Returns the first inequality filter contained within this composite filter. Returns null if it
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java:55]: `s/null/{@code null}`

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java (hunk 17 above)
Comment (schmidt-sebastian) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java:47]: Please add TODO, otherwise we will never look at this code again.

**Review by schmidt-sebastian — APPROVED** [review]

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java (hunk 21 above)
Comment (ehsannas) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java]: Done

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java (hunk 17 above)
Comment (ehsannas) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/core/CompositeFilter.java:48]: Done
